File: donuts-factory-pact/src/test/java/com/bil/contract/donuts/ContractTest.java
package com.bil.contract.donuts;

import au.com.dius.pact.consumer.Pact;
import au.com.dius.pact.consumer.PactProviderRuleMk2;
import au.com.dius.pact.consumer.PactVerification;
import au.com.dius.pact.consumer.dsl.PactDslWithProvider;
import au.com.dius.pact.model.RequestResponsePact;
import feign.Feign;
import feign.Response;
import feign.gson.GsonEncoder;
import org.junit.Rule;
import org.junit.Test;

import java.util.HashMap;
import java.util.Map;

import static org.junit.Assert.assertEquals;

public class ContractTest {


    DonutsClient donutsClient = Feign.builder().encoder(new GsonEncoder())
            .target(DonutsClient.class, "http://localhost:8085");

    @Rule
    public PactProviderRuleMk2 mockProvider =
            new PactProviderRuleMk2("donuts-api-pact", "localhost", 8085, this);


    @Pact(consumer="donuts-factory-pact")
    public RequestResponsePact donutsOk(PactDslWithProvider builder) throws Exception {
        return builder
                .given("")
                .uponReceiving("Represents a successful scenario of creating a donut")
                .path("/donuts")
                .method("POST")
                .headers("Content-Type", "application/json")
                .body("{\"butter\": 120, \"sugar\": 150, \"flour\": 200}")
                .willRespondWith()
                .status(201)
                .headers(responseHeaders())
                .given("")
                .uponReceiving("Represents a successful scenario of getting all donuts")
                .path("/donuts")
                .method("GET")
                .willRespondWith()
                .status(200)
                .body("[{\"butter\": 120, \"sugar\": 150, \"flour\": 200}]")
                .headers(responseHeaders())
                .toPact();
    }

    @Test
    @PactVerification
    public void runDonuts() {
        Response response = this.donutsClient.createDonuts(new Donut(120, 150, 200));
        assertEquals(response.status(), 201);

        response = this.donutsClient.getDonuts();
        assertEquals(response.status(), 200);

    }

    private Map<String, String> responseHeaders() {
        Map<String, String> map = new HashMap<>();
        map.put("Content-Type", "application/json;charset=UTF-8");
        return map;
    }

}

File: simpson-client-pact/src/test/kotlin/contract/simpson/ContractTest.kt
package contract.simpson

import au.com.dius.pact.consumer.MockServer
import au.com.dius.pact.consumer.dsl.PactDslJsonArray
import au.com.dius.pact.consumer.dsl.PactDslJsonBody
import au.com.dius.pact.consumer.dsl.PactDslWithProvider
import au.com.dius.pact.consumer.junit5.PactConsumerTestExt
import au.com.dius.pact.consumer.junit5.PactTestFor
import au.com.dius.pact.core.model.RequestResponsePact
import au.com.dius.pact.core.model.annotations.Pact
import com.bil.contract.simpson.Donut
import com.bil.contract.simpson.DonutsClient
import feign.Feign
import feign.gson.GsonEncoder
import org.junit.jupiter.api.extension.ExtendWith
import kotlin.test.Test
import kotlin.test.assertEquals

@ExtendWith(PactConsumerTestExt::class)
class ContractTest {

    @Pact(provider = "donuts-factory-pact", consumer = "simpson-pact")
    fun `create donuts`(builder: PactDslWithProvider): RequestResponsePact {
        val donutsBody = PactDslJsonArray().minArrayLike(1)
                .integerType("butter")
                .integerType("sugar")
                .integerType("flour")
                .closeObject()

        val donutBody = PactDslJsonBody()
                .integerType("butter")
                .integerType("sugar")
                .integerType("flour")

        return builder
                .given("creating successfully a donut")
                .uponReceiving("all ingredients for a donut")
                    .path("/donuts")
                    .method("POST")
                    .headers("Content-Type", "application/json")
                    .body(donutBody)
                .willRespondWith()
                    .status(201)
                .given("failing to create a donut")
                .uponReceiving("some missing ingredients")
                    .path("/donuts")
                    .method("POST")
                    .headers("Content-Type", "application/json")
                    .body("{\"butter\":-100, \"sugar\": 150, \"flour\": 200}")
                .willRespondWith()
                    .status(400)
                .given("getting all donuts")
                .uponReceiving("every produced donuts by the factory")
                    .path("/donuts")
                    .method("GET")
                .willRespondWith()
                    .status(200)
                    .body(donutsBody)
                    .headers(mapOf("Content-Type" to "application/json"))
                .toPact()
    }

    @Test
    @PactTestFor(pactMethod = "create donuts")
    fun `check create donuts`(mockServer: MockServer) {
        val donutsClient = Feign.builder()
                .encoder(GsonEncoder())
                .target(DonutsClient::class.java, mockServer.getUrl())

        assertEquals(201, donutsClient.createDonuts(Donut(120, 150, 200)).status())

        assertEquals(400, donutsClient.createDonuts(Donut(-100, 150, 200)).status())

        assertEquals(200, donutsClient.getDonuts().status())

    }
}
